### ISSUE TITLE
require-await should allow no await if returning a promise and other items may return/throw prior

### ISSUE THREAD

**Issue by ArcticZeroo** [eslint/eslint#10462]
**What rule do you want to change?**
require-await

**Does this change cause the rule to produce more or fewer warnings?**
fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option, called "with-promise-return" (or similar, give me a break on thinking of a good name -- it's midnight)

The option will not cause the rule to flag if this condition is met:
- [ ] The async function returns a promise
And one or more of the following conditions are met:
- [ ] The async function has the ability to return prior to returning the promise
- [ ] The async function has the ability to throw an error prior to returning the promise

**Please provide some example code that this change will affect:**

```js
    /**
     * Subscribe to user presence
     * for all cached users.
     * @return {Promise}
     */
    async sendUserPresenceSub() {
        const readyState = this.api.rtm.socket.readyState;

        if (readyState < ReadyState.OPEN) {
            return new Promise((resolve ,reject) => {
                this.api.rtm.once('open', () => this.sendUserPresenceSub().then(resolve).catch(reject));
            });
        }

        if (readyState > ReadyState.OPEN) {
            throw new Error('WebSocket was closed, could not subscribe to presence');
        }

        return new Promise((resolve, reject) => {
            this.api.rtm.socket.send(
                JSON.stringify({
                    type: 'presence_sub',
                    ids: Array.from(this.users.keys())
                }),
                function (err) {
                    if (err) {
                        return reject(err);
                    }

                    resolve();
                }
            );
        });
```
Don't worry about the semantics of what's going on in this code, I can find better snippets later if need be, but for the time being this is the actual code flagging the error for me right now.

**What does the rule currently do for this code?**
It flags it as being invalid because await is required for all async functions.

**What will the rule do after it's changed?**
It will allow this function because the first, and at least one of the following two conditions are met
- [x] The async function returns a promise
- [x] The async function has the ability to return prior to returning the promise
- [x] The async function has the ability to throw an error prior to returning the promise

This is important because we would otherwise need to wrap the whole thing in an unnecessary additional promise to get the same functionality without an async function. Of course, async functions treat returns as resolves, throws as rejects, and returning a promise absorbs the resolve/reject behavior.

**Comment by ljharb** [comment]
How can eslint statically know if it's returning a promise?

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
